### ISSUE TITLE
Couple of Bugs

### ISSUE THREAD

**Issue by Maxetto** [Bimmr/McInfected#72]
Hi,
I'm testing this plugin, however in the last version v1.2.9 (stated as v1.3.0 in plugin.yml) there are a couple of Bugs, some are Major, others are just minimal.

1) Ranks' Permissions don't update at the end of the game. I have disabled Ranks Permissions by default, and I'm using Ranks to update their Kit by removing the old Kit permission and giving the new one. People can't use the new Kit even if set as Default by the rank until I reload the plugin.

1b) It seems to issue the Rank permissions [twice](http://pastebin.com/1VT5rexJ).

2) Killstreaks doesn't work. In console I see "[Unknown Command.](http://pastebin.com/bDsNgntD)" when a KillStreak should be issued.

3) Couple of errors in 1.7 Protocol Hack (with Carbon) regarding [COOKED_MUTTON](http://pastebin.com/8cRB9dHd) when opening the Human Kit Menu and [FoodHealer](http://pastebin.com/i7QmGRZk) when interacting with items.

4) Using "Book" item to vote for a Map doesn't seem to work on 1.8.8 (while it was working fine on 1.7.10 Protocol Hack)

5) Kit selection and Map voting messages (errors too) are sometime issued twice.

6) Missing CraftPlayer cast on commands (aka. Issuing commands on console gives [error](http://pastebin.com/vyMAUXmc)).

7) McInfected-Dedicated messages can't be translated (they are reset to default at every reload/restart)

8) --This is a feature request-- Configurable grenades with effects, like in the older Infected (I don't know why they were removed...)

Thank you for this awesome plugin, hope to see it getting better and better over time.

Using:
PaperSpigot 1.7.10 (Protocol Hack w/ Carbon) - Spigot 1.8.8
McInfected v1.2.9
McInfected-Ranks v1.1.6
McInfected-Dedicated v0.1.3
McInfected-LobbyItems v0.0.7
Vault v1.4.1 - v1.5.6
PermissionsEx v1.22.10 - v1.23.4

**Comment by Bimmr** [comment]
Glad to see that someone knows how to make a bug ticket :laughing:, I've had way to many people just make a ticket that says "This doesn't work" then doesn't post any more info.

I'll start looking into them. 

2. Can I see your setup for that killstreak(So I can see exactly how the command is written)

4. Are you using a book or a book and quill? On my test server I have it running Spigot 1.8.8 and the book and quill work fine. 

5. Not to sure what you mean? Do the messages display more than once?

8. I decided to let Crackshot handle grenades because that plugin was coded with guns and grenades in mind as where McInfected wasn't.

**Comment by Maxetto** [comment]
1) Sure, here it is: [http://pastebin.com/mbKXNcHi](http://pastebin.com/mbKXNcHi)

2) I was using WRITTEN_BOOK (as it was McInfected-LobbyItems default, I think), on 1.7 it was working fine, in 1.8 it opens the book showing "Invalid book Tag **". Not sure if it's an EssentialsX issue though.

3) Yes, messages are displayed twice.

4) Then, is it possible to use Crackshot without guns (just grenades)?

**Comment by Bimmr** [comment]
1. For a quick fix change "item:" to "id:".
I will fix it in the next version.

2. Normally when I see that invalid book tag it was when I was playing on my friends server and I had a written book on a 1.8 server that had just recently updated from 1.7.10

4. You can use "gun:<grenade name>" to give a class a grenade

**Comment by Maxetto** [comment]
1) Fixed, thanks.
2) Yes, it's exactly that. I updated to 1.8.8 from 1.7.10. I'll use the "EMPTY_MAP" item, no big deal ;)

As for Rank permissions I think I found the problem. Permissions are given in the world, instead of being global.

**Comment by Maxetto** [comment]
Other Bugs:
9) It seems like CrackShot grenades don't do anything, not damage players nor apply potion effects. I don't know what causes this, I've tried disabling WorldGuard (which is the only world protection plugin I use) with no luck.

10) When giving compass to zombies in the last 60 seconds, when they die, the compass is not given anymore. Is this intended?

**Comment by Bimmr** [comment]
9. It may be a bug with crackshot but I'll look into it. 

10. The compass is meant to be just a ping to show their last known location.

11. I don't think there is a type modifier for items, you're best bet is just to make a custom potion

12. I am in the process of adding an unbreakable modifier.

13. I'll make it so golden apples heal you to full because that sounds like a good idea, but I'm not able to make it give the effects because of the way the food healing system works.

**Comment by Maxetto** [comment]
13) That's what I was meaning, but a Bug that could occur is that if you're Full Health, you can eat the Golden Apple normally and cause that to give you the effects.

**Comment by Bimmr** [comment]
Just a list of everything in this ticket

- [x] Use Global permissions instead of per world
- [x] Killstreaks aren't working for the tag "item:"
- [x] Older versions of Spigot don't have Cooked Mutton in the material enum
- [ ] Books don't work with Carbon trying to update from 1.7.10 to 1.8
- [ ] Kit and map selection messages appear twice
- [x] Admin commands don't work from console
- [ ] McInfected-Dedicated's messages reset on restart
- [ ] Crackshot grenades don't work 
- [x] Add an unbreakable modifier
- [x] Golden Apples & God Apples heal to max health

**Comment by Maxetto** [comment]
14) In a BungeeCord environment, joining a server when there's less than the minimum required users to start the game and/or (not tested) in Lobby, doesn't seem to teleport the players to the Lobby Spawnpoint.

Tested both with and without OP/Permissions. May also be caused by AuthMe.

**Comment by Bimmr** [comment]
Does it teleport the player at all or do they just stay at spawn?

**Comment by Maxetto** [comment]
They spawn where they left the last time. (Also if I leave from another World)

**Comment by Bimmr** [comment]
Are they using a command to leave?

**Comment by Maxetto** [comment]
Nevermind, leaving infected was the "bug".
No issues occurs if you leave when you're inside Infected. :wink:

**Comment by Bimmr** [comment]
While looking at the Crackshot bug i remembered I made a ticket on crackshot. So it looks like the author hasn't fixed the grenade bug. http://dev.bukkit.org/bukkit-plugins/crackshot/tickets/294-weapon-damage-entity-event-and-explosion_no_damage/

**Comment by Maxetto** [comment]
Another thing, this plugin seems to edit fall damage, which causes NoCheatPlus to behave strangely.
Expecially if you jump on a water block (only 1 layer, nothing under) from a certain height (2-3, don't remember), you'll get pushed back in the water, allowing you to swim without issues (while instead you should fall under the map and die).

Using: NoCheatPlus [#928](http://ci.md-5.net/job/NoCheatPlus/928/)

**Comment by Bimmr** [comment]
I dont edit fall damage done to players.

**Comment by Maxetto** [comment]
- [ ] 21) Regeneration Potion doesn't regen Health.
- [ ] 24) --Feature Request-- Configurable Wait Time before Game Start (after voting phase) and after Game End (before teleporting back to Lobby).
- [x] 25) --Feature Request-- Transfer Potions to self and allies and, if not already, possibility to transfer more than 1 potion after hit (Like [this](http://pastebin.com/fQN2uwsm))
- [x] 27) --Feature Request-- Option to make KillStreaks for Infecteds stay after death. (As it's not likely that Infecteds can get high KillStreaks compared to Humans)
- [x] 29) First Spawn spawns everyone in only one spawn even if more than 1 are configured.

[Kits.yml](http://pastebin.com/WWAVvTjZ)
- [x] EDIT: This is not really a Bug with McInfected, but I suggest you to set Ranks varchar(20) to something like varchar(30) or more. I had to set it to 25 because one of my rank had 24 characters.

**Comment by Maxetto** [comment]
Solved the Grenade issue by myself. I had my world in Peaceful Difficulty, which didn't apply explosion damages...

**Comment by Maxetto** [comment]
32) It's impossible to choose Baby Disguises with Lib's Disguises.

**Comment by Bimmr** [comment]
20) Why do you have multiple potions in a single slot anyways...
31) I don't edit the damage an arrow does, so It might be another plugin
32) I don't really see why you'd want to disguise as a baby though....

**Comment by Maxetto** [comment]
18) Error Logs are reachable from "crashes" and "error" words.
20) Probably comfort, I didn't want potions to fill the Hotbar.
31) Will investigate more...
32) High Speed, Low Damage and Armor Baby Zombie Kit.

**Comment by Bimmr** [comment]
18) Sorry, had something enabled in my web browser which made links not look like links...

**Comment by Maxetto** [comment]
Ticket cleaned and updated.

Many of the new Bugs (such as Lib's Disguises compatibility) are absolutely Major issues.

**Comment by Bimmr** [comment]
Add the tag "unbreakable" to an item, and i completely forgot about the voting bug

**Comment by Bimmr** [comment]
45) Is a YAML limitation, try \: or /:

**Comment by Maxetto** [comment]
45) Doesn't work neither with \:, /:, \u003a, &#58;.

**Comment by Maxetto** [comment]
BimmCore 0.0.4 [completely](http://pastebin.com/vFg5hR0M) breaks SQL statements.

**Comment by Bimmr** [comment]
MySQL is one of the things i forgot to test before uploading, turns out SQL was completely broken for one of McInfected's addons aswell.

**Comment by Maxetto** [comment]
Ticket updated again.
Stated some priorities:
~~Strikethrough~~: Tested and fixed
**Bold**: High Priority, Game Breaking issue.
_Italic_: Normal Priority, issues that needs to be fixed but are not that important.
Normal Text: Low Priority, Awaiting confirmation, low priority issues or feature requests.

**Comment by Bimmr** [comment]
I seem to be able to set the icon for the kits, as well as armor seems to be working for me.

**Comment by Maxetto** [comment]
I can set the Icon for Kits too, but after I load the plugin, they get away.

I think the same thing happens to Armor too.

**Comment by Bimmr** [comment]
http://img.prntscr.com/img?url=http://i.imgur.com/zVhVSI4.png
![image](https://cloud.githubusercontent.com/assets/5053870/15445913/5ab8893c-1eda-11e6-907d-e495f89ef3f3.png)

The icons are still there after the server has been restarted a bunch of times.

**Comment by Bimmr** [comment]
Only thing I noticed is that the glow doesn't show.

**Comment by Maxetto** [comment]
After the conversion the Kits.yml breaks and removes "Icon:", "Helmet:", "Chestplate:", "Leggings:" and "Boots:" lines.
Glow works fine for me.
Armor and Icons works until you restart or reload the plugin.

May it be caused by Spigot version or one of yours extensions (Versions used are written on the first post)?

**Comment by Bimmr** [comment]
Double check that all the addons are the latest version, as well as BimmrCore and McInfected. I'm not able to get the icons/armor to break.

**Comment by Bimmr** [comment]
Nevermind, the armor just stopped working... odd...

**Comment by Bimmr** [comment]
Try removing LibsDisguise off the server then tell me if armor is still broken, because it looks like LibsDisguise will remove the armor, but still apply the protection when hit.

**Comment by Maxetto** [comment]
Besides that Lib's Disguises isn't working with latest versions of McInfected, it's not a problem with it because armor and icon datas are removed from the Kits.yml

**Comment by Bimmr** [comment]
They aren't being removed from the file for me. So I really don't know what
to do/say, and just to double check, you are using the latest mcInfected
and addons as well as bimmcore?

**Comment by Maxetto** [comment]
I confirm the versions stated on the first post and I confirmed they are the latest possible to get on 1.8.X

**Comment by Maxetto** [comment]
Still waiting for at least this fix, I can't even run the test server since the game is broken :confused: 
[ ] 34) Hook with Lib's Disguises doesn't work with older version of Lib's Disguises (for those still in 1.8.8).

**Comment by Bimmr** [comment]
I am working on the updates in my free time, because I do not make enough
to put a focus on them. But recently I've ran into an issue with my
computer that I am trying to fix

**Comment by Maxetto** [comment]
Updated:
Added bug 49 and 50, promoted bug 48 to high priority, updated bug 21 and 29 with more information, marked bug 33 as fixed after some testing.

**Comment by DanielMac1** [comment]
17) --Feature Request-- Zombie spawned at game start gets a different Kit than the default one if he has not selected another. I want that too!

**Comment by Bimmr** [comment]
@Maxetto Since you seem to be good at finding bugs in my plugins, would you want to test the new version? I haven't uploaded it because I need to test it still but It's always easier to break it when you get someone else to test it.

**Comment by DanielMac1** [comment]
Ill test it i found this voting bug when there is more than 16 maps it breaks.

**Comment by DanielMac1** [comment]
Ill test it i found this voting bug when there is more than 16 maps it breaks.

**Comment by Maxetto** [comment]
@Bimmr Of course, I'm always up for developer/beta versions. However I'm studying for my final exams, so expect slow or even no updates until 4 July.

**Comment by DanielMac1** [comment]
Pm me a download :palm_tree:

**Comment by Bimmr** [comment]
Went to send you both a download then noticed I still had a few things to fix... Just finally getting around to it, so hopefully I'll be able to send the files in a day or two

**Comment by Maxetto** [comment]
I finished my final exams and I'm finally on Vacation :stuck_out_tongue_closed_eyes:
I'm ready to test almost 24/7.

**Comment by Bimmr** [comment]
K, Currently I'm just having trouble finding time to fix a bug that shouldn't be happening at all in the first place (Like all bugs....)

**Comment by Maxetto** [comment]
> The bug forces the software to adapt, evolve into something new because of it. Work around it or work through it. No matter what, it changes. It becomes something new. The next version. The inevitable upgrade.
- Mr. Robot

**Comment by Maxetto** [comment]
Here are some new bugs from the Testing version:
- [x] ~~Arena name in Scoreboard have "_" instead of Space~~
- [x] ~~and trimms with more than 11 characters (due to the Spacing before name).~~
- [x] ~~Changes done to armor don't save inside the Kits.yml file (When you reload/restart the server, changes are gone).~~
- [ ] Player names that have an underscore ("_") inside get replaced with space :stuck_out_tongue_closed_eyes: 
- [x] ~~Title messages are printed twice/seems glitched.~~
- [ ] _Baby Zombie disguise still not works (Disguise: Zombie baby). (Not working on iDisguises, throws error)_
- [x] ~~In Lobby and in Pre-Game you can "die" by falling.~~

Confirmed fixed bugs are striked on the first message. Old fixed bugs are removed.

**Comment by Bimmr** [comment]
I just realised I sent the BimmCore that had all my Title/BossBar/Scoreboard testing code in it...

**Comment by Maxetto** [comment]
Feature Request #27 seems implemented in the dev version but It doesn't seem configurable (or at least I didn't check)

**Comment by Bimmr** [comment]
For the baby disguises, what disguise plugin are you using?

**Comment by Maxetto** [comment]
For the Disguises I'm using Lib's Disguises v8.6.8 (for Minecraft 1.8.8/9).

I edit files directly, I never use In-Game configurators.

**Comment by Bimmr** [comment]
Do you edit the files when the server is running or when it's stopped?

**Comment by Maxetto** [comment]
Tried both, same result.

When Infected loads the Kits.yml, it change it by removing (but loading correctly) armor settings like "unbreakable", "enchant", etc. I thought when shutting down the server, the file would be saved correctly again (due to the load/save thing I saw in BimmCore code), but I wasn't right.
It's like the old issue where Armor and Icon settings weren't saved at all, but now it's only for their settings.

**Comment by Bimmr** [comment]
If it was missing the Unbreakable, Flags, and attributes I think i'd know the issue. But if it's also missing enchantment then i'll have to mess around with it.

**Comment by Maxetto** [comment]
I have tried changing enchantment from full name (like "ENVIRONMENT_PROTECTION") to ID (like "0") and it does not save.

May be caused because there's "unbreakable" tag too. Will try and report in a minute.

**Comment by Bimmr** [comment]
So it's only changing from "enchantment:0" to "enchantment:PROTECTION"?

**Comment by Maxetto** [comment]
Yeah, it's going back to "enchantment:PROTECTION". It does that with every armor piece, but not items.

**Comment by Bimmr** [comment]
K, It will no longer change things for armor.

**Comment by Bimmr** [comment]
With MySQL, does it still save and load to the database? Because the falling back message looks like it just can't find a debugger class.

**Comment by Maxetto** [comment]
Yes, it's still working. But it falls back to something that looks like less efficient.

**Comment by Maxetto** [comment]
Updated. Most bugs where fixed, I split the Arena Name in Scoreboard bug in 2 as it was fixed the "_"/" " bug but not the 12+ character trimm.
Added a bug where a Debug message is shown also with "Debug: false"

**Comment by Bimmr** [comment]
I've been trying to get the PlayerToggleSneakEvent error but nothing is causing the error to pop up. Do you have the error?

**Comment by Maxetto** [comment]
I have grouped all of them [HERE](http://pastebin.com/HjmUAE6k), is the same Link as it is in the word "Error".

**Comment by Maxetto** [comment]
Confirmed fixed the Name bug by latest dev build.

**Comment by Bimmr** [comment]
Awesome, anything else major that you noticed?

**Comment by Maxetto** [comment]
Nothing else seems to break the game, but there are some not-game breaking bugs that are important (the ones in Italic)

**Comment by Maxetto** [comment]
Fun fact:
With the new version, besides that granades deals damage (or at least 1 one my type), eating a Golden Apple while full life, makes an ANGRY_VILLAGER sound, but eating isn't cancelled.
